src/main/java/com/avanza/gs/test/junit5/ChainableBeforeCallbacks.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright 2017 Avanza Bank AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.avanza.gs.test.junit5;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.junit.jupiter.api.extension.BeforeAllCallback;
+import org.junit.jupiter.api.extension.BeforeEachCallback;
+import org.junit.jupiter.api.extension.ExtensionContext;
+
+interface ChainableBeforeCallbacks extends BeforeAllCallback, BeforeEachCallback {
+
+	default ChainableBeforeCallbacks andThen(ChainableBeforeCallbacks next) {
```

Comment:
This method does not seem to be used (other than from unit tests). Can we remove it or move it to the unit-tests that need it?

Reply:
No, this construction is a replacement for JUnit 4's `RuleChain`. It is only used in tests currently because this is a test library. Modules that use gs-test and want to replace JUnit 4 with JUnit 5 can chain these resources together using this mechanism.

Reply:
Does this replace the `@Order` annotation on extensions when used by tests? When would a caller decide to use `.andThen` compared to specifying two extensions with different `@Order`? Would this interface need to be `public` for a caller to be able to use it, or is it only intended for instances of PUs that need to start in a particular order?

Reply:
Yes, this can be used instead of `@Order`.
A caller would choose to use the `andThen` mechanism when replacing an existing `RuleChain` because the structure is more similar. 
`@Order`would be the standard JUnit 5 way of doing this, the `andThen` mechanism is intended to ease transition from JUnit 4 and `RuleChain`
The interface does not need to be public.

Here is an example of how it could be used, note that this is a different package from the `ChainableBeforeCallbacks` interface:
```java
package com.avanza.gs.test.junit5.test;

import org.junit.jupiter.api.extension.Extension;
import org.junit.jupiter.api.extension.RegisterExtension;

import com.avanza.gs.test.junit5.PuConfigurers;
import com.avanza.gs.test.junit5.RunningPu;

public class MyTest {

	private final RunningPu pu = PuConfigurers.partitionedPu("pu.xml").configure();
	private final RunningPu mirrorPu = PuConfigurers.mirrorPu("mirror-pu.xml").configure();

	@RegisterExtension
	Extension extension = pu.andThen(mirrorPu);

}
```

Reply:
So an implementing app would change its existing junit4 test from:
```java
final RunningPu pu = ...;
final RunningPu mirrorPu = ...;

@Rule
public RuleChain chain = RuleChain.outerRule(pu).around(mirrorPu);
```

to
```java
final RunningPu pu = ...;
final RunningPu mirrorPu = ...;

@RegisterExtension
Extension extension = pu.andThen(mirrorPu);
```

or (the standard junit5 way):
```java
@RegisterExtension
final RunningPu pu = ...;

@RegisterExtension
final RunningPu mirrorPu = ...;
```

If you feel that `.andThen` is easier to understand for developers compared to the standard way then lets go with that.

Reply:
Yes! However the standard JUnit 5 way would have to be
```java
@Order(1)
@RegisterExtension
final RunningPu pu = ...;

@Order(2)
@RegisterExtension
final RunningPu mirrorPu = ...;
```
Because otherwise the order is not guaranteed.

I actually prefer using the standard JUnit 5 way, but I figured this alternative would make the transition easier. Maybe it would be better to just recommend using the standard way though.

As a sidenote, `RuleChain` in JUnit 4 should not be used for ordering rules either, according to the Javadoc:

> In older versions of JUnit (before 4.13) RuleChain was used for ordering rules. We recommend to not use it for this purpose anymore. You can use the attribute order of the annotation Rule or ClassRule for ordering rules.

Reply:
Good. Sure, it's up to you if we want to provide a different way than the standard junit-way, if that makes things easier for implementing apps. Otherwise I suggest we close the suggestion provided in this conversation thread.

Reply:
I removed the `andThen` functionality and updated the test to use `@Order`

Reply:
👍